src/Language/DifferentialDatalog/DatalogProgram.hs

```diff
@@ -307,3 +311,221 @@ progMirrorInputRelations d prefix =
     rules = map (\(n,r) -> makeRule n r) inputRels
   in d { progRelations = M.union (progRelations d) $ M.fromList relCopies,
          progRules     = (progRules d) ++ rules }
+
+-- For RHSLiteral, a binding to the expression is inserted if it's not bound to a variable.
+-- For example, R(a, b, z, _) gets transformed into __r0 in R(a, b, z, _),
+addBindingToRHSLiteral :: (RuleRHS, Int) -> RuleRHS
```

Comment:
One more comment I missed in my initial review: I think all the logic related to injecting debugging hooks deserves a separate module: `src/Language/DifferentialDatalog/Debug.hs`.

Reply:
Done.